scripts/deploy.py

```diff
@@ -43,18 +43,68 @@ def upload_archive(archive, zenodo_json, version):
     headers = {"Accept": "application/json"}
     params = {"access_token": ZENODO_TOKEN}
 
-    # Create an empty upload
-    response = requests.post(
-        "https://zenodo.org/api/deposit/depositions",
-        params=params,
-        json={},
-        headers=headers,
-    )
-    if response.status_code not in [200, 201]:
-        sys.exit(
-            "Trouble requesting new upload: %s, %s"
-            % (response.status_code, response.json())
+    if doi:
+        depositions = requests.get(
+            "https://zenodo.org/api/deposit/depositions",
+            params=params,
+            headers=headers,
+        )
+        if depositions.status_code not in [200, 201]:
+            sys.exit("Cannot query depositions: %s, %s" % (depositions.status_code, depositions.json()))
+
+        target_deposit = None
+        for deposit in depositions.json():
+            print("looking at deposit %s" % (deposit['doi']))
+            if deposit['conceptdoi'] == doi:
+                target_deposit = deposit
+
+        if not target_deposit:
+            sys.exit("Cannot find deposit with doi: '%s'. Are you currently editing it?" % (doi))
+
+        # found the existing deposit - so let's make a new version.
+        url = "%s/actions/newversion" % target_deposit['links']['self']
+        new_version = requests.post(
+            url,
+            params=params,
+            headers=headers,
+        )
+        if new_version.status_code not in [200, 201]:
+            sys.exit("Cannot create a new version for doi '%s'. %s" % (doi, new_version.json()))
+
+        # new_version is actually the next draft. cannot edit the existing doi above
+        new_version = requests.get(
+            new_version.json()['links']['latest_draft'],
+            params=params,
+            headers=headers,
+        )
+        if new_version.status_code not in [200, 201]:
+            sys.exit("Cannot create a new version for doi '%s'. %s" % (doi, new_version.json()))
+
+        # this draft is based off of version N-1, so let's remove N-1's artifacts to make room
+        # for version N.
+        for file in new_version.json()['files']:
```

Comment:
Could you show me what it looks like to delete and not delete old versions? I'm not convinced that we want to do this, especially if someone has referenced something via the DOI. The whole sort of philosophy around an archive is that it's stable so this makes me uncomfortable.

Reply:
https://zenodo.org/record/6325907 is where I'd tripped up on this.

I was trying to release `2.0.23`, but it still has the zip file for `2.0.22`. If I were to keep going then `N+100` version would have all 99 previous versions in it. So I'm deleting the previous version's files while in this draft state - effectively creating a new empty deposit like the else block.

> This action will create a new deposit, which will be a snapshot of the current resouce, inheriting the metadata as well as snapshot of files. The new version deposit will have a state similar to a new, unpublished deposit, most importantly its files will be modifiable as for a new deposit.

The documentation around `newversion`. You can't edit files of published artifacts. I don't even think you can edit _anything_ once you've published.

Reply:
Oh interesting. So for the action I think we have a few options:

1. Make the "cleanup" a variable, giving the user the choice to delete or not. I'd make the default to not delete, only because it's the more conservative action, but I would also be okay with having cleanup true the default if you think that is the more common desire. The issue with cleanup is that we lose the previous versions, and that's probematic for reproducibility. There are likely projects out there that aren't going to release as often as ondemand, OR they just want their uses to have access to previous versions. Let Zenodo deal with the issue of too many files :)  Let me know your thoughts.

2. Make it more an all or nothing thing - if you want to maintain the same DOI, you get the older ones deleted. If you want a new doi, you get a new doi with the new version. I'm not a fan of this one because I'd like the choice to use a common DOI and have artifacts there. Option 1 above is a much better UI for the user to have a choice!

3. Allow the user to specify a third variable (or possibly more) to keep versions based on some criteria. E.g., you could imagine something like "keep the last 10" or "keep the last version of a major release" or "keep all major releases." I like that option too.

If you want to decide on the basic functionality (without the extra params) and then get the staging/test working here, I'd be happy to follow up with a PR or two this evening / later this week! I recognize that this is way out of scope for the work you are fixing here, so I'm happy to help.

Reply:
My leaning is toward #1. I'd be happy with a default of false, but I would actually guess (naively) that true is the more common case. Especially if the previous versions are immutable with their own unique DOIs.

I'm not a researcher, so I'm really just the producer of DOIs and not really a consumer looking to use it as a citation so I could be off base. It just seems like a `1:1` relationship is straight forward.  

That said I don't feel that strongly about it, because as an option I can just change it, so it's all good.

Reply:
@johrstrom to step back a second - I'm wondering why it's not possible to release multiple versions associated with the same DOI, e.g., ideally we'd have this outcome: https://zenodo.org/record/6310818. Each different version is associated with the DOI (and we don't need to delete anything, another version is just added).

Reply:
oh I see - are you saying that when you create a new version, it sort of clones the original and includes the previous file? I'm looking at your example and you indeed have these multiple versions:

![image](https://user-images.githubusercontent.com/814322/157126170-969df804-6f68-4a97-90b6-21c3977a7da5.png)

Reply:
Sure it's possible. I don't _need_ to delete the files - I just wanted to to make it look clean.

We're just uploading the 1 zip - the one file and it has a version in it, so Zenodo let's us just append the new file to the filelist. Version `2.0.23` of ondemand is what this looks like - https://zenodo.org/record/6325907 - you just keep adding a `.zip` to a file list.  But the next version, 2.0.23-1-1 https://zenodo.org/record/6325946, is more of what I'm looking for.

It just didn't make sense to me to provide a citation for `2.0.23` (a DOI with versions 22 and 23) when the DOI for 2.0.22 itself exists and is a part of the same DOI group.

Actually that parent DOI is a good example of what would happen if you flipped that flag back and forth between releases.
```
22 - 1 file
23 - two files (newest appended to existing list)
23-1-1 one file
```

Reply:
That feels buggy to me - it should not be the case that the new version associated with a DOI inherits the previous files. Do you mind if I pull this branch and play around? I can probably do it later this evening (not during "work time" lol).

Reply:
Sure - I'm done I think, though I may open some tickets while it's on my mind.

If you do merge it, can you release it? I'll update ondemand to start using it.

Reply:
yes of course! I'll probably be opening another PC.

I have some other open source work on my queue first (and currently having dinner) but I expect to have time tonight to work on this.